ansible_base/rbac/api/views.py

```diff
@@ -0,0 +1,99 @@
+from ansible_base.rbac.models import RoleDefinition
+
+
+class RoleDefinitionViewSet(ModelViewSet, AnsibleBaseDjangoAppApiView):
+    """
+    Role Definitions (roles) contain a list of permissions and can be used to
+    assign those permissions to a user or team through the respective
+    assignment endpoints.
+
+    Custom roles can be created, modified, and deleted through this endpoint.
+    System-managed roles are shown here, which cannot be edited or deleted,
+    but can be assigned to users.
+    """
+
+    queryset = RoleDefinition.objects.all()
+    serializer_class = RoleDefinitionSerializer
+    permission_classes = [AuthenticatedReadAdminChange]
+
+    def get_serializer_class(self):
+        if self.action == 'update':
+            return RoleDefinitionDetailSeraizler
+        return super().get_serializer_class()
+
+    def _error_if_managed(self, instance):
+        if instance.managed is True:
+            raise ValidationError('Role is managed by the system')
+
+    def perform_update(self, serializer):
+        self._error_if_managed(serializer.instance)
+        return super().perform_update(serializer)
+
+    def perform_destroy(self, instance):
+        self._error_if_managed(instance)
+        return super().perform_destroy(instance)
+
+
+class BaseAssignmentViewSet(ModelViewSet, AnsibleBaseDjangoAppApiView):
+    permission_classes = [permissions.IsAuthenticated]
+    # PUT and PATCH are not allowed because these are immutable
+    http_method_names = ['get', 'post', 'head', 'options', 'delete']
+
+    def get_queryset(self):
+        model = self.serializer_class.Meta.model
+        if has_super_permission(self.request.user, 'view'):
+            return model.objects.all()
+        return model.visible_items(self.request.user).prefetch_related('content_object')
+
+    def perform_destroy(self, instance):
+        if not self.request.user.has_obj_perm(instance, 'delete'):
+            raise PermissionDenied
+
+        rd = instance.object_role.role_definition
+        obj = instance.object_role.content_object
+        with transaction.atomic():
+            rd.remove_permission(self.request.user, obj)
+
+
+class RoleTeamAssignmentViewSet(BaseAssignmentViewSet):
+    """
+    Use this endpoint to give a user permission to a resource or an organization.
+    The needed data is the user, the role definition, and the object id.
+    The object must be of the type specified in the role definition.
+    The type given in the role definition and the provided object_id are used
+    to look up the resource.
+
+    After creation, the assignment cannot be edited, but can be deleted to
+    remove those permissions.
+    """
+
+    serializer_class = RoleTeamAssignmentSerializer
+
+
+class RoleUserAssignmentViewSet(BaseAssignmentViewSet):
+    """
+    Use this endpoint to give a team permission to a resource or an organization.
+    The needed data is the team, the role definition, and the object id.
```

Comment:
docstrings seem to be swapped with team, right?

Reply:
Yep.

I should provide some context on why a 2-line class has a paragraph docstring. This text will get rendered in the browsable API. So this is directly user/client facing. Probably good to get the model right.